truffle/src/com.oracle.truffle.api.strings.test/src/com/oracle/truffle/api/strings/test/taint/SwitchEncodingTests.java

```diff
@@ -0,0 +1,85 @@
+package com.oracle.truffle.api.strings.test.taint;
+
+import com.oracle.truffle.api.strings.AbstractTruffleString;
+import com.oracle.truffle.api.strings.TruffleString;
+import org.junit.Test;
+
+import static com.oracle.truffle.api.strings.test.taint.TaintTestUtils.convert;
+import static com.oracle.truffle.api.strings.test.taint.TaintTestUtils.from;
+import static com.oracle.truffle.api.strings.test.taint.TaintTestUtils.getTaint;
+import static com.oracle.truffle.api.strings.test.taint.TaintTestUtils.isTainted;
+import static com.oracle.truffle.api.strings.test.taint.TaintTestUtils.length;
+import static com.oracle.truffle.api.strings.test.taint.TaintTestUtils.taint;
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class SwitchEncodingTests {
+
+    @Test
+    public void switchingEncodingShouldPersistTaintAscii() {
+        for (TruffleString.Encoding source : TruffleString.Encoding.values()) {
+            AbstractTruffleString a = taint(from("foo", source), "bar");
+            Object[] taintArrSource = getTaint(a);
+            assertTrue("Source should be tainted", isTainted(a));
+            for (TruffleString.Encoding target : TruffleString.Encoding.values()) {
+                AbstractTruffleString b = convert(a, target);
+                assertTrue(
+                        String.format("Target should be tainted (%s -> %s)", source, target),
+                        isTainted(b)
+                );
+                Object[] taintArrTarget = getTaint(b);
+                assertArrayEquals(
+                        String.format("Taint labels should be the same (%s -> %s)", source, target),
+                        taintArrSource,
+                        taintArrTarget
+                );
+            }
+        }
+    }
+
+    @Test
+    public void switchingEncodingShouldPersistTaintNonAscii() {
```

Comment:
UTF8_MAC changes "foo" into some 5 char length string, I have skipped this encoding for now.